chain/ethereum/src/network.rs

```diff
@@ -48,64 +72,129 @@ impl EthereumNetworkAdapter {
+    ) -> Self {
+        use std::cmp::Ordering;
+
+        use graph::slog::{o, Discard, Logger};
+
+        use graph::components::adapter::MockIdentValidator;
+        let chain_id: ChainId = "testing".into();
+        adapters.sort_by(|a, b| {
+            a.capabilities
+                .partial_cmp(&b.capabilities)
+                .unwrap_or(Ordering::Equal)
+        });
+
+        let provider = ProviderManager::new(
+            Logger::root(Discard, o!()),
+            vec![(chain_id.clone(), adapters)].into_iter(),
+            Arc::new(MockIdentValidator),
+        );
+        provider.mark_all_valid().await;
+
+        Self::new(chain_id, provider, call_only, None)
+    }
+
+    pub fn new(
+        chain_id: ChainId,
+        manager: ProviderManager<EthereumNetworkAdapter>,
+        call_only_adapters: Vec<EthereumNetworkAdapter>,
+        retest_percent: Option<f64>,
+    ) -> Self {
+        #[cfg(debug_assertions)]
+        call_only_adapters.iter().for_each(|a| {
+            a.is_call_only();
+        });
+
+        Self {
+            chain_id,
+            manager,
+            call_only_adapters,
+            retest_percent: retest_percent.unwrap_or(DEFAULT_ADAPTER_ERROR_RETEST_PERCENT),
         }
     }
-    pub fn all_cheapest_with(
-        &self,
+
+    fn available_with_capabilities<'a>(
+        input: Vec<&'a EthereumNetworkAdapter>,
         required_capabilities: &NodeCapabilities,
-    ) -> impl Iterator<Item = &EthereumNetworkAdapter> + '_ {
-        let cheapest_sufficient_capability = self
-            .adapters
+    ) -> impl Iterator<Item = &'a EthereumNetworkAdapter> + 'a {
+        let cheapest_sufficient_capability = input
             .iter()
             .find(|adapter| &adapter.capabilities >= required_capabilities)
             .map(|adapter| &adapter.capabilities);
 
-        self.adapters
-            .iter()
+        input
+            .into_iter()
             .filter(move |adapter| Some(&adapter.capabilities) == cheapest_sufficient_capability)
             .filter(|adapter| adapter.get_capacity() > AvailableCapacity::Unavailable)
     }
 
-    pub fn cheapest_with(
+    pub async fn all_cheapest_with(
         &self,
         required_capabilities: &NodeCapabilities,
+    ) -> impl Iterator<Item = &EthereumNetworkAdapter> + '_ {
+        let all = self
+            .manager
+            .get_all(&self.chain_id)
+            .await
+            .unwrap_or_default();
+
+        Self::available_with_capabilities(all, required_capabilities)
+    }
+
+    // get all the adapters, don't trigger the ProviderManager's validations because we want
+    // this function to remain sync
+    pub(crate) fn all_unverified_cheapest_with(
+        &self,
+        required_capabilities: &NodeCapabilities,
+    ) -> impl Iterator<Item = &EthereumNetworkAdapter> + '_ {
+        let all = self
+            .manager
+            .get_all_unverified(&self.chain_id)
+            .unwrap_or_default();
```

Comment:
Same comment on the `unwrap_or_default` as above

Reply:
same